csharp/src/Google.Protobuf/Reflection/Dynamic/DynamicMessage.cs

```diff
@@ -0,0 +1,377 @@
+﻿using Google.Protobuf.Collections;
+using System;
+using System.Collections;
+
+namespace Google.Protobuf.Reflection.Dynamic
+{
+
+    /// <summary>
+    /// An implementation of IMessage that can represent arbitrary types, given a MessageaDescriptor.
+    /// Unknown fields not supported.
+    /// </summary>
+    public sealed partial class DynamicMessage : IMessage
+    {
+        private readonly MessageDescriptor type;
+        private readonly FieldSet fields;
+        private int memoizedSize = -1;
+
+        /// <summary>
+        /// Descriptor for this message.
+        /// </summary>
+        public MessageDescriptor Descriptor => type;
+
+        /// <summary>
+        /// Default implementation of Overridden method.
+        /// </summary>
+        /// <param name="input"></param>
+        /// <exception cref="NotImplementedException"></exception>
+        public void MergeFrom(CodedInputStream input)
+        {
+            throw new NotImplementedException();
```

Comment:
Is there any built-in functionality that depends on this? What doesn't work due to this not being implemented? FWIW looking at the Java implementation as reference, this is in place there.

Reply:
tried to keep this implementation as close to java implementation as possible. This specific implementation is never invoked. MergeFrom functionality comes into picture when we do deserialization from ByteString. Deserialization is triggered by invoking ParseFrom static method of DynamicMessage, which calls the mergeFrom of MessageExtension, which further invokes mergeFrom of DynamicMessage.Builder.
Updated the message comments with details.